### PR TITLE
ROX-14660: Fix overwrites to node scans from Node pipeline

### PR DESCRIPTION
**Pull request by jvdm**
## Description

Add an allow list to match Node's `OsImage` attributes and determine if they support full host-level node scanning. If yes:

1. Ignore the node during reprocessing.
2. Handle `SensorEvent_Node` by not enriching the node and skipping risk calculation.
3. Handle `SensorEvent_NodeInventory` by enriching the inventory and attaching a `Node.Scan`, leaving the other metadata intact.

The above ensures both pipelines won't overwrite each other data as long as they don't run concurrently. `SensorEvent_Node` should only update Node, but not `Node.Scan`, while `SensorEvent_NodeInventory` should only update `Node.Scan` but nothing else.

For simplicity, we will continue to allow these pipelines to perform reads before writes to the database to update only parts that are relevant for each. And we will ensure they don't run concurrently by using the same [worker queue when processing them](https://github.com/stackrox/stackrox/pull/5123).

To ensure `Node` is not touched by `SensorEvent_NodeInventory`, we keep `Node.LastUpdated` if already set on upsert. Tests were updated to account for the fact.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] ~Evaluated and added CHANGELOG entry if required~ N/A
- [ ] ~Determined and documented upgrade steps~ N/A
- [ ] ~Documented user-facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~ N/A

## Testing Performed

Integration tests were added to cover the scenario where `Node` is being upserted with `Node.Scan = nil`.

For manual testing, deploy to Openshift-4.12 (cluster `jvdm-argilo`). and:

1. Reduce reprocessing time: `kubectl -n stackrox set env deployment/central ROX_REPROCESSING_INTERVAL=1m`
1. Reduce node scanning time: `kubectl -n stackrox set env daemonset/collector "ROX_NODE_SCANNING_INTERVAL_DEVIATION=30s" "ROX_NODE_SCANNING_INTERVAL=1m" "ROX_NODE_SCANNING_INTERVAL_MAX_INITIAL_WAIT=0s"`
1. Two experiments:
   a. Run reprocessing and inventories at a high rate (~1m), and trigger node events by kicking Sensor.
   a. Disable inventories with `kubectl -n stackrox set env daemonset/collector "ROX_RHCOS_NODE_SCANNING=false"` and trigger node events.

### UI

image

### Logs

Reprocessing:

```
reprocessor: 2023/03/10 01:09:13.069413 reprocessor.go:420: Info: Successfully reprocessed 87/87 images
reprocessor: 2023/03/10 01:09:13.069609 reprocessor.go:421: Info: Resyncing deployments now that images have been reprocessed...
reprocessor: 2023/03/10 01:09:35.625355 reprocessor.go:273: Info: Found 6 nodes to scan
reprocessor: 2023/03/10 01:09:35.648246 reprocessor.go:445: Info: node secured-cluster-jvdm-argilo:jvdm-argilo-47bkl-master-2.c.srox-temp-dev-test.internal (id: c3d2395f-ed74-455f-b9f6-4507db02c779) does not n
reprocessor: 2023/03/10 01:09:35.650914 reprocessor.go:445: Info: node secured-cluster-jvdm-argilo:jvdm-argilo-47bkl-master-0.c.srox-temp-dev-test.internal (id: 1ce86640-8407-4e01-aed6-665da2362936) does not n
reprocessor: 2023/03/10 01:09:35.652714 reprocessor.go:445: Info: node secured-cluster-jvdm-argilo:jvdm-argilo-47bkl-master-1.c.srox-temp-dev-test.internal (id: 3c995dde-9890-48de-94e9-93bffdce2876) does not n
reprocessor: 2023/03/10 01:09:35.654843 reprocessor.go:445: Info: node secured-cluster-jvdm-argilo:jvdm-argilo-47bkl-worker-b-h7xp9.c.srox-temp-dev-test.internal (id: ea130298-20a1-4bbc-a7ac-04a00a0a21e3) does
reprocessor: 2023/03/10 01:09:35.655099 reprocessor.go:445: Info: node secured-cluster-jvdm-argilo:jvdm-argilo-47bkl-worker-c-6ps8s.c.srox-temp-dev-test.internal (id: 9105af0f-522f-4381-a0f2-39399263e52e) does
reprocessor: 2023/03/10 01:09:35.668693 reprocessor.go:445: Info: node secured-cluster-jvdm-argilo:jvdm-argilo-47bkl-worker-d-lkv4x.c.srox-temp-dev-test.internal (id: 3c73854f-9852-4de8-8a57-d5230c8cfde5) does
reprocessor: 2023/03/10 01:09:35.668803 reprocessor.go:302: Info: Successfully reprocessed 0/6 nodes
```

NodeInventory:

```
sensor/service/pipeline/nodeinventory: 2023/03/10 01:08:02.357826 pipeline.go:65: Info: Received NodeInventory for Node name='jvdm-argilo-47bkl-worker-c-6ps8s.c.srox-temp-dev-test.internal' ID='9105af0f-522f-4381-a0f2-39399263e52e'
sensor/service/pipeline/nodeinventory: 2023/03/10 01:08:07.812502 pipeline.go:65: Info: Received NodeInventory for Node name='jvdm-argilo-47bkl-master-2.c.srox-temp-dev-test.internal' ID='c3d2395f-ed74-455f-b9f6-4507db02c779'
sensor/service/pipeline/nodeinventory: 2023/03/10 01:11:23.420295 pipeline.go:65: Info: Received NodeInventory for Node name='jvdm-argilo-47bkl-master-0.c.srox-temp-dev-test.internal' ID='1ce86640-8407-4e01-aed6-665da2362936'
sensor/service/pipeline/nodeinventory: 2023/03/10 01:11:55.890659 pipeline.go:65: Info: Received NodeInventory for Node name='jvdm-argilo-47bkl-worker-b-h7xp9.c.srox-temp-dev-test.internal' ID='ea130298-20a1-4bbc-a7ac-04a00a0a21e3'
sensor/service/pipeline/nodeinventory: 2023/03/10 01:12:17.636049 pipeline.go:65: Info: Received NodeInventory for Node name='jvdm-argilo-47bkl-master-1.c.srox-temp-dev-test.internal' ID='3c995dde-9890-48de-94e9-93bffdce2876'
sensor/service/pipeline/nodeinventory: 2023/03/10 01:12:22.904669 pipeline.go:65: Info: Received NodeInventory for Node name='jvdm-argilo-47bkl-worker-d-lkv4x.c.srox-temp-dev-test.internal' ID='3c73854f-9852-4de8-8a57-d5230c8cfde5'
```

## Multi-cluster test

In a multi-cluster setup, one with RHCOS scanning *enabled* the other with RHCOS *disabled*:

image

### CODE REVIEW

File: central/node/datastore/store/postgres/store.go
Comment (vikin91) [inline central/node/datastore/store/postgres/store.go:506]: Help me understand this - why we should still do the upsert if there are no changes in both: the metadata and scan?

File: central/reprocessor/reprocessor_unit_test.go
Comment (vikin91) [inline central/reprocessor/reprocessor_unit_test.go:43]:
```suggestion
			name: "when OS is not RHCOS, scanner is called and node is upserted",
```

File: central/reprocessor/reprocessor_unit_test.go
Comment (vikin91) [inline central/reprocessor/reprocessor_unit_test.go:61]:
```suggestion
			name: "when get node returns err, reprocessNode returns false",
```
or
```suggestion
			name: "false should be returned when get node returns err",
```

File: central/reprocessor/reprocessor_unit_test.go
Comment (vikin91) [inline central/reprocessor/reprocessor_unit_test.go:78]: Finish is optional here - see https://pkg.go.dev/github.com/golang/mock/gomock#NewController
We do not call finish in `central/sensor/service/pipeline/nodeinventory/pipeline_test.go`, so maybe let's pick one variant (calling it or not calling it) and apply globally.

File: central/sensor/service/pipeline/nodeinventory/pipeline_test.go
Comment (vikin91) [inline central/sensor/service/pipeline/nodeinventory/pipeline_test.go:48]: It looks like this could be a bool in this test, as the only thing we do with that is checking whether this is empty `tt.wantErr != ""`.

As another option, we could assert using:

```go 
s.ErrorContains(err, tt.wantErr)
```

**Review by vikin91 — COMMENTED** [review]
This looks good and I found mostly nits, but the main point that I would like understand is why the changes to `isUpdated` are needed - i.e., why we no longer need to know whether the node metadata was updated?

File: central/node/datastore/store/postgres/store.go
Comment (jvdm) [reply central/node/datastore/store/postgres/store.go:506]: Previously, `metadataUpdated` was always set to `true`, meaning this if the statement was redundant.

File: central/reprocessor/reprocessor_unit_test.go
Comment (jvdm) [reply central/reprocessor/reprocessor_unit_test.go]: Fixed at 751aa60dc2

File: central/reprocessor/reprocessor_unit_test.go
Comment (jvdm) [reply central/reprocessor/reprocessor_unit_test.go]: Thanks! Quoting docs:

> New in go1.14+, if you are passing a *testing.T into NewController function you no longer need to call ctrl.Finish() in your test methods. 

Fixed at 751aa60dc2

File: central/node/datastore/store/postgres/store.go
Comment (vikin91) [reply central/node/datastore/store/postgres/store.go:506]: I see. Shouldn't then we change 

```go
	if !metadataUpdated && !scanUpdated {
		return nil
	}
```

into

```go
	if !scanUpdated {
		return nil
	}
```
?

File: central/sensor/service/pipeline/nodeinventory/pipeline_test.go
Comment (jvdm) [reply central/sensor/service/pipeline/nodeinventory/pipeline_test.go:48]: At [central/sensor/service/pipeline/nodeinventory/pipeline_test.go:53](https://github.com/stackrox/stackrox/pull/5145/files/9765951442ead153bc3455ae5590ba18b7af0a3f#diff-323ec1a5481b6d118c6a4ba909e66baa7f3f7fc469525cfb4bcb564abefd2a80R53) we check the error's text.

Comment (vikin91) [reply central/sensor/service/pipeline/nodeinventory/pipeline_test.go:49]: I see that we do set it on line 53, but the only place we read it is in https://github.com/stackrox/stackrox/pull/5145/files/9765951442ead153bc3455ae5590ba18b7af0a3f#diff-323ec1a5481b6d118c6a4ba909e66baa7f3f7fc469525cfb4bcb564abefd2a80R95 where we only compare it against `""`.

File: central/node/datastore/store/postgres/store.go
Comment (jvdm) [reply central/node/datastore/store/postgres/store.go:506]: Why?

Comment (vikin91) [reply central/node/datastore/store/postgres/store.go:506]: I see that the if-clause would never evaluate to true and we would never quit early here. However, I am thinking about the intentions of the original author of that code, and I think the following was meant: "if there were no changes to metadata and no changes to scan data, then quit early". By removing the entire if, we do keep the same behavior but we dismiss the original intention to quit early if there are no changes. Is there any value in quitting early `if !scanUpdated`?

**Review by vikin91 — APPROVED** [review]
It looks good to me! 
I am not fluent with that part of the test code, so I didn't do a thorough review of that part (but I really like the `TestStore_UpsertWithoutScan`), so maybe consider getting another review before merging? (@md2119 is the ideal choice 😄 )

File: central/node/datastore/store/postgres/store.go
Comment (jvdm) [reply central/node/datastore/store/postgres/store.go:506]: There is no value in quitting early `if !scanUpdate`. In fact, that is one of the cases we expect to occur when on `SensorEvent_Node`. The pipeline will update the Node with outdated (in fact `nil`) scan data. If we give up on that scenario, we prevent the update altogether. Unless there is a scenario in which both `last_updated` and `scan_time` are older, I am not aware of any.

File: central/reprocessor/reprocessor.go
Comment (md2119) [inline central/reprocessor/reprocessor.go:445]: Consider `cluster/node` convention which is followed in other places as well as UI.

File: central/reprocessor/reprocessor.go
Comment (md2119) [inline central/reprocessor/reprocessor.go:457]: We should avoid setting fields controlled by store elsewhere. Please take a look at my other comment about last updated.

File: central/reprocessor/reprocessor.go
Comment (md2119) [inline central/reprocessor/reprocessor.go:445]: What does non-legacy node mean? If we skip enrichment, how do we make ensure the snoozed state of vulns is updated?

File: central/node/datastore/store/postgres/store.go
Comment (vikin91) [reply central/node/datastore/store/postgres/store.go:506]: @md2119 scans can come now from two sources: nodes pipeline (aka V1 scan) & nodeInventory pipeline (aka V2 scan). The latter has richer set of components and vulnerabilities, but is reported every 4 hours. The Node pipeline has fewer scan data but is reported on every node update from Sensor. We do not want that I fresher V1 scan replaces older but still valid (and reacher) V2.

Is that what was meant here?

Comment (md2119) [reply central/node/datastore/store/postgres/store.go:506]: If that is a guarantee, can we move the onus from the low-level store to a separate reconciler? Both node pipelines feed into this reconciler. When central receives node scans from either of the pipelines, the reconciler forwards or drops the message based on what you stated above. Therefore, the reconciler becomes the only channel/entry point from the datastore perspective. This way we move the complexity out of the low-level store.

Comment (vikin91) [reply central/node/datastore/store/postgres/store.go:506]: @jvdm what do you think about that idea?

Comment (jvdm) [reply central/node/datastore/store/postgres/store.go:506]: @md2119:

> I am not following the discussion entirely. Why would the an outdated scan have scan time later than recent scan time?

The `SensorEvent_Node` is sent when there are updates in the Kubernetes Node resource. That event used to be the only node event. And we would use the metadata (by that, I mean orchestrator, kernel, etc.) from the Node to generate scan components.

With Host-Level Node Scanning, we will not use the metadata from the `SensorEvent_Node` to generate scan components. Instead, with Host-Level Node Scanning, we have a new event, `SensorEvent_NodeInventory`, that arrives at every configurable interval (default ~4hs) with the scan components. We still need to handle `SensorEvent_Node`, but we don't need to populate the `Node.Scan` anymore when handling that event. That is only done when handling `SensorEvent_NodeInventory`.

That's the scenario where `Node.Scan` would be "outdated": When handling `SensorEvent_Node`. When writing the node to the store when handling `SensorEvent_Node`, the scan data should be ignored and pulled from the DB (if there is one). On the other hand, the other event `SensorEvent_NodeInventory` should not touch the Node metadata.

File: central/node/datastore/datastore.go
Comment (jvdm) [inline central/node/datastore/datastore.go:111]: I've added this here because my initial feeling is that this logic should be close to its model, and I could not find a better candidate. Please correct me if you know a better place.

**Comment by jvdm** [comment]
I reverted the `LastUpdated` to allow the field. I intuitively assumed they should be independent, but I don't think it is worth discussing this considering how minimal the impact on customers is, and also @md2119 is advocating keeping the datastore "dump" and "decisionless", which is an agreed goal.

The options considered were, selected (c.):

(a.) Do a Get before calling Upsert.
(b.) Add "options" to Upsert APIs to change its behavior and either overwrite or pull fields before writing.
(c.) Check `Node.Scan == nil` and perform a Get inside the datastore (selected :heavy_check_mark:).

File: central/node/datastore/datastore.go
Comment (md2119) [reply central/node/datastore/datastore.go:111]: `pkg/node`

File: central/sensor/service/pipeline/nodes/pipeline.go
Comment (md2119) [inline central/sensor/service/pipeline/nodes/pipeline.go:97]: I want to confirm my thought about the `if` block:
You are skipping the enrichment and analysis for host-scannable nodes because the inventory pipeline is guaranteed to provide more accurate vuln insights.

If the above is true, it would be a good idea to include it in the comment.

File: central/sensor/service/pipeline/nodes/pipeline.go
Comment (md2119) [inline central/sensor/service/pipeline/nodes/pipeline.go:98]: super minor nit: `upserting`

File: central/sensor/service/pipeline/nodes/pipeline.go
Comment (md2119) [reply central/sensor/service/pipeline/nodes/pipeline.go:97]: Additionally, I think you mentioned that you need to handle `SensorEvent_Node` for metadata but don't need to populate the `.scan` from this pipeline. How about setting it .`.scan` to `nil` here? That should guarantee the scan from this pipeline would never disturb scan in db because of the check `scan == nil` in store.

File: central/sensor/service/pipeline/nodeinventory/pipeline.go
Comment (md2119) [inline central/sensor/service/pipeline/nodeinventory/pipeline.go:98]: By now, we have the correct metadata and scan because metadata is pulled from db, that is given by old pipeline, whereas the scan is provided by the inventory pipeline, correct?

File: central/node/datastore/store/postgres/store_test.go
Comment (md2119) [inline central/node/datastore/store/postgres/store_test.go:139]: Check if nil scan overwrites scan in db.

**Review by md2119 — COMMENTED** [review]
Dropped a couple of comments, otherwise lgtm. One comment overall, consider putting the pipeline changes behind the feature flag.

File: central/sensor/service/pipeline/nodeinventory/pipeline.go
Comment (jvdm) [reply central/sensor/service/pipeline/nodeinventory/pipeline.go:98]: Correct.

File: central/node/datastore/store/postgres/store_test.go
Comment (md2119) [reply central/node/datastore/store/postgres/store_test.go:139]: Ignore, my oversight

File: central/sensor/service/pipeline/nodes/pipeline.go
Comment (jvdm) [reply central/sensor/service/pipeline/nodes/pipeline.go:107]: > I see that you would like to communicate more to the users

I want to clarify what this warning means for Customers, Support, and Engineers.

> The error and continued processing?

Yes, both. By ignoring the error, I continue the processing.

> How about something that informs the effect? e.g. risk analysis of the node maybe inaccurate?

Yeah, why not. But "maybe" is not helpful, to be honest. But we can be specific: "enriching node %s failed, the failure was ignored: vulnerability and risk information will not be updated."

File: central/sensor/service/pipeline/nodeinventory/pipeline.go
Comment (md2119) [inline central/sensor/service/pipeline/nodeinventory/pipeline.go:77]: Here, if you are pulling the node only for the metadata, then you can use `GetNodeMetadata`. It is more performant because no cycles are spent on pulling scan data.

Comment (jvdm) [reply central/sensor/service/pipeline/nodeinventory/pipeline.go:77]: But if enriching fails, later on, I will overwrite with an empty scan. I leave it as is.

Comment (jvdm) [reply central/sensor/service/pipeline/nodeinventory/pipeline.go:77]: I will follow up on this in a separate PR. I want to add a single API to retrieve one single NodeMetadata instead of using GetMany().

Comment (jvdm) [reply central/sensor/service/pipeline/nodeinventory/pipeline.go:77]: Also need to double-check what metadata brings in and see if we are not missing any key piece of information. And write tests around it.

Comment (md2119) [reply central/sensor/service/pipeline/nodeinventory/pipeline.go:77]: Sounds like a reasonable follow-up action.

File: central/sensor/service/pipeline/nodes/pipeline.go
Comment (md2119) [reply central/sensor/service/pipeline/nodes/pipeline.go:99]: > Maybe. I don't think there is ambiguity here regarding the role of the two pipelines for host-scannable nodes, are you feeling strong about this?

It is a change in behavior with the launch of this feature, and not many people are aware about it. It is just to save time explaining to individual basis.

> .scan is already always nil. Or, do you mean explicitly setting it to nil, to guard the beavior? That is a good idea. Let me know if you meant something else.

Correct, I meant extra safety valve.

**Review by md2119 — APPROVED** [review]

**Pull request merged**
